### PR TITLE
added a --clock-source argument, defaulting to "internal"

### PR DESCRIPTION
**Pull request by marcusmueller**
so that uhd_fft can use an external clock source. Time source doesn't make much sense in the context of the apps using this.

### CODE REVIEW

**Comment by jmcorgan** [comment]
@mbr0wn Please look at this and label it ready for merge if you're happy with it.

**Comment by mbr0wn** [comment]
@marcusmueller you're potentially overwriting UHD defaults. `uhd_fft` should *not* provide a default, and then
````
    if args.clock_source is not None: set_clock_source(args.clock_source)
````

**Comment by marcusmueller** [comment]
Ah, right.

**Comment by mbr0wn** [comment]
@marcusmueller Are you going to update this branch?

**Comment by marcusmueller** [comment]
Yes, but not before Tuesday


Am 21. Januar 2016 20:08:41 GMT+00:00, schrieb mbr0wn <notifications@github.com>:
>@marcusmueller Are you going to update this branch?
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/gnuradio/gnuradio/pull/719#issuecomment-173693783

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by mbr0wn** [comment]
That's alright, I just wanted to know if you were owning this task. Thanks!

M

**Comment by marcusmueller** [comment]
fixed.

**Comment by mbr0wn** [comment]
Sorry, I still have complaints...

This code should try and do as little as possible that's also done in UHD. Your code pre-defines valid options for the clock source, which a) depend on the device configuration and b) might not even be a complete list in some future version of UHD.

So, just skip the pre-defined options, and pass it straight through to UHD. If it fails, it'll have a clear error message. Sure, it'll fail a bit later but at least the code doesn't imply anything (e.g. this code implies that selecting gpsdo is valid, but often enough it's not).

I'll add specific inline comments.

File: gr-uhd/apps/uhd_app.py [hunk 1]
```diff
@@ -295,5 +298,7 @@ def cslist(string):
                             help="Show asynchronous message notifications from UHD")
         group.add_argument("--sync", choices=('default', 'pps', 'auto'),
                           default='auto', help="Set to 'pps' to sync devices to PPS")
+        group.add_argument("--clock-source",choices=('internal', 'external', 'gpsdo', 'default'),
```
Comment (mbr0wn) [inline gr-uhd/apps/uhd_app.py]: Simply remove the choices here. However, you can add them in the help string ("Set the clock source (e.g., internal, external, gpsdo, ...)").

File: gr-uhd/apps/uhd_app.py [hunk 2]
```diff
@@ -141,6 +141,9 @@ def setup_usrp(self, ctor, args, cpu_format='fc32'):
         if args.spec:
             for mb_idx in xrange(self.usrp.get_num_mboards()):
                 self.usrp.set_subdev_spec(args.spec, mb_idx)
+        # Set the clock source:
+        if args.clock_source and not args.clock_source == "default":
```
Comment (mbr0wn) [inline gr-uhd/apps/uhd_app.py]: Here, simply do
```{.py}
if args.clock_source is not None:
    self.usrp.set_clock_source(...)
```

**Comment by marcusmueller** [comment]
As much as I share that sentiment (do not duplicate UHD functionality),
UHD examples do that all over the place (and it's ugly); I think the
original motivation was to "fail early" rather than letting the user
wait until the device is fully initialized, just to inform him that our
devices only support these three options.

However, more than happy to do as you suggested, because if I hadn't
worked with UHD examples, I'd never have the idea to offer but a fixed
set of options.

**Comment by marcusmueller** [comment]
done :)

**Comment by mbr0wn** [comment]
@jmcorgan ready for merge.

**Comment by jmcorgan** [comment]
This was squashed as a single commit to master.

**Pull request closed without merging**